inc/azure_c_util/thandle.h

```diff
@@ -0,0 +1,189 @@
+
+/*given a previous type T introduced by MU_DEFINE_STRUCT(T, T_FIELDS), this is the name of the type that has T wrapped*/
+#define THANDLE_WRAPPER_TYPE_NAME(T) MU_C2(T, _WRAPPER)
+
+/*given a previous type T, THANDLE_MALLOC introduces a new name that mimics "malloc for T"*/
+/*the new name is used to define the name of a static function that allocates memory*/
+#define THANDLE_MALLOC(T) MU_C2(T,_MALLOC)
+
+/*given a previous type T, THANDLE_FREE introduces a new name that mimics "free for T"*/
+/*the new name is used to define the name of a static function that free memory allocated by THANDLE_MALLOC*/
+#define THANDLE_FREE(T) MU_C2(T,_FREE)
+
+/*given a previous type T, THANDLE_DEC_REF introduces a new name that mimics "dec_ref for T"*/
+/*the new name is used to define the name of a static function that decrements the ref count of T and if 0, releases it*/
+#define THANDLE_DEC_REF(T) MU_C2(T,_DEC_REF)
+
+/*given a previous type T, THANDLE_INC_REF introduces a new name that mimics "inc_ref for T"*/
+/*the new name is used to define the name of a static function that increments the ref count of T*/
+#define THANDLE_INC_REF(T) MU_C2(T,_INC_REF)
+
+/*given a previous type T, THANDLE_ASSIGN introduces a new name for a function that does T1=T2 (with inc/dec refs)*/
+#define THANDLE_ASSIGN(T) MU_C2(T,_ASSIGN)
+
+/*given a previous type T, THANDLE_INITIALIZE introduces a new name for a function that does T1=T2 (with inc ref, and considers T1 as uninitialized memory)*/
+#define THANDLE_INITIALIZE(T) MU_C2(T,_INITIALIZE)
+
+/*given a previous type T introduced by MU_DEFINE_STRUCT(T, T_FIELDS), this introduces:
+1. the type that wraps it
+2. THANDLE_MALLOC macro to create it, initialize refCount to 0, and remember the dispose function*/
+
+#define THANDLE_TYPE_DEFINE(T) \
+    MU_DEFINE_STRUCT(THANDLE_WRAPPER_TYPE_NAME(T), T, data, THANDLE_EXTRA_FIELDS(T));                                                                                   \
+    static T* THANDLE_MALLOC(T)(void(*dispose)(T*))                                                                                                                     \
+    {                                                                                                                                                                   \
+        T* result;                                                                                                                                                      \
+        /*Codes_SRS_THANDLE_02_013: [ THANDLE_MALLOC shall allocate memory. ]*/                                                                                         \
+        THANDLE_WRAPPER_TYPE_NAME(T)* handle_impl = (THANDLE_WRAPPER_TYPE_NAME(T)*)malloc(sizeof(THANDLE_WRAPPER_TYPE_NAME(T)));                                        \
+        if (handle_impl == NULL)                                                                                                                                        \
```

Comment:
These macro that we write start to look like templates :-)

Reply:
"If the compiler doesn't do it!!! We have to do it ourselves!!!"

---
In reply to: [385373033](https://github.com/Azure/azure-c-util/pull/16#discussion_r385373033) [](ancestors = 385373033)